internal/apis/certmanager/validation/certificaterequest.go

```diff
@@ -191,98 +208,128 @@ func ValidateUpdateCertificateRequestApprovalCondition(oldCRConds, newCRConds []
+	// CertificateRequest.
+
+	if hasExtension(csr, pki.OIDExtensionBasicConstraints) {
+		if !template.BasicConstraintsValid {
+			return fmt.Errorf("BasicConstraintsValid %v does not match expected value %v", template.BasicConstraintsValid, true)
 		}
-	}
 
-	return newUsages
-}
+		if template.IsCA != isCA {
+			return fmt.Errorf("IsCA %v does not match expected value %v", template.IsCA, isCA)
+		}
 
-func isUsageEqual(a, b []cmapi.KeyUsage) bool {
-	a = patchDuplicateKeyUsage(a)
-	b = patchDuplicateKeyUsage(b)
+		expectedMaxPathLen := 0
+		expectedMaxPathLenZero := false
+		if maxPathLen != nil {
+			template.MaxPathLen = *maxPathLen
+			template.MaxPathLenZero = *maxPathLen == 0
+		}
 
-	var aStrings, bStrings []string
+		if template.MaxPathLen != expectedMaxPathLen {
+			return fmt.Errorf("MaxPathLen %v does not match expected value %v", template.MaxPathLen, expectedMaxPathLen)
+		}
 
-	for _, usage := range a {
-		aStrings = append(aStrings, string(usage))
+		if template.MaxPathLenZero != expectedMaxPathLenZero {
+			return fmt.Errorf("MaxPathLenZero %v does not match expected value %v", template.MaxPathLenZero, expectedMaxPathLenZero)
+		}
 	}
 
-	for _, usage := range b {
-		bStrings = append(bStrings, string(usage))
+	// We validate that the provided keyUsage and extKeyUsage match the
+	// usages that are present in the certificate.
+
+	if hasExtension(csr, pki.OIDExtensionKeyUsage) {
+		if keyUsage != template.KeyUsage {
+			return fmt.Errorf("the KeyUsages %v do not match the expected KeyUsages %v",
+				printKeyUsage(apiutil.KeyUsageStrings(template.KeyUsage)),
+				printKeyUsage(apiutil.KeyUsageStrings(keyUsage)),
+			)
+		}
 	}
 
-	return util.EqualUnsorted(aStrings, bStrings)
+	if hasExtension(csr, pki.OIDExtensionExtendedKeyUsage) {
+		if !equalExtKeyUsage(extKeyUsage, template.ExtKeyUsage) {
+			return fmt.Errorf("the ExtKeyUsages %v do not match the expected ExtKeyUsages %v",
+				printKeyUsage(apiutil.ExtKeyUsageStrings(template.ExtKeyUsage)),
+				printKeyUsage(apiutil.ExtKeyUsageStrings(extKeyUsage)),
+			)
+		}
+	}
+
+	return nil
+}
+
+type printKeyUsage []cmapiv1.KeyUsage
+
+func (k printKeyUsage) String() string {
+	var sb strings.Builder
+	sb.WriteString("[")
+	for i, u := range k {
+		sb.WriteString(" '")
+		sb.WriteString(string(u))
+		sb.WriteString("'")
+		if i < len(k)-1 {
+			sb.WriteString(",")
+		}
+	}
+	if len(k) > 0 {
+		sb.WriteString(" ")
+	}
+	sb.WriteString("]")
+	return sb.String()
 }
 
-// ensureCertSignIsSet adds UsageCertSign in case it is not set
-// TODO: add a mutating webhook to make sure this is always set
-// when isCA is true.
-func ensureCertSignIsSet(list []cmapi.KeyUsage) []cmapi.KeyUsage {
-	for _, usage := range list {
-		if usage == cmapi.UsageCertSign {
-			return list
+func equalExtKeyUsage(a, b []x509.ExtKeyUsage) bool {
+	if len(a) != len(b) {
+		return false
+	}
+
+OuterLoop:
+	for _, va := range a {
+		for _, vb := range b {
+			if va == vb {
+				continue OuterLoop
+			}
```

Comment:
I don't think we are importing golang.org/x/exp/slices already. Maybe we should wait until this is part of the std, and prevent adding this dependency now?

Reply:
I can't really imagine a safer dependency to add, despite the `exp` in the name! I'd vote to add it because:

1) We know it's going into the stdlib soonish so we know it's temporary
2) It's basically stdlib anyway
3) It means I / we don't need to think about the correctness of this function or whether it's optimal or not.

On (3), what I mean is that it looks to me that this function could be something like:


```go
for i := 0; i < 10; i++ {
	if a[i] != b[i] { return false }
}

return true
```

But I'd rather just not have to think about that, you know?